### PR TITLE
Anand dev heatmap yolo: current branch with changes

### PR DESCRIPTION
**Pull request by AnandSingh-0619**
## Changelog

1. Policy: [YoloSAMPointNavResNetPolicy](https://github.com/AnandSingh-0619/home-robot/blob/anand_dev_heatmap_yolo/projects/habitat_uncertainty/models/YOLOSAMPointNavResNetPolicy.py)

- Created [YOLOPerception class](https://github.com/AnandSingh-0619/home-robot/blob/anand_dev_heatmap_yolo/projects/habitat_uncertainty/models/YOLOSAMPointNavResNetPolicy.py#L212C1-L342C30) 
__init__ : initializes YOLO and MobileSAM models and 
predict: give object and receptacle segmentation masks as outputs. Also added some filters to reduce number of loop executions containing call to the SAM predict function.

- Modified forward function of the [GazeResNetEncoder](https://github.com/AnandSingh-0619/home-robot/blob/anand_dev_heatmap_yolo/projects/habitat_uncertainty/models/YOLOSAMPointNavResNetPolicy.py#L454C1-L477C45)
__init__: removed head_rgb from the visual encoder visual keys. Also initializes the YOLO_perception class
forward: It first checks if the "yolo_segmentation_sensor" is in observations (which is always the case as the collect_env_res from trainer calls predict function of YOLOPerception and stores that output) and then modifies that data according to the current skill (here GAZE: object_segmentation and start_recep_segmentation)

- Modified [GazePointNavResNetNet](https://github.com/AnandSingh-0619/home-robot/blob/anand_dev_heatmap_yolo/projects/habitat_uncertainty/models/YOLOSAMPointNavResNetPolicy.py#L676C1-L676C194)
__init__: removed head_rgb, yolo_object_sensor, yolo_start_receptacle_sensor and yolo_segmentation_sensor from fuse_keys

2. Trainer: [YOLOSAMTrainer](https://github.com/AnandSingh-0619/home-robot/blob/anand_dev_heatmap_yolo/projects/habitat_uncertainty/trainers/YOLOSAM_trainer.py)
_init_train: Called the YOLOPerception predict function to return masks and store it as reading of the yolo_segmentation_Sensor for the first_rollout. Also remove extra observation keys: head_rgb etc.
 _collect_environment_result: Make call to the YOLOPerception predict function and then clean the batch observation to store for the rollout_storage.
 train: added some extra print statements for debugging and clear cache memory.

3. Sensors: [sensor](https://github.com/AnandSingh-0619/home-robot/blob/anand_dev_heatmap_yolo/projects/habitat_uncertainty/task/sensors.py)
yolo_object_sensor: returns target object  category id as per new mapping of classes
yolo_start_receptacle_sensor: returns start receptacle id
yolo_goal_receptacle_sensor: returns goal receptacle id
yolo_segmentation_sensor: returns dummy tensor of same size as output for YOLOPerception predict function.

4. Single agent access Manager: [single_agent_access_manager](https://github.com/AnandSingh-0619/home-robot/blob/anand_dev_heatmap_yolo/projects/habitat_uncertainty/models/single_agent_access_manager.py)
_create_storage: removes extra keys from stoage from observation space-head_rgb, yolo_start_receptacle_sensor, yolo_object_sensor
load_state_dict: Fixed the previous error of lr_lambda resume state

5. Config: [config.py](https://github.com/AnandSingh-0619/home-robot/blob/anand_dev_heatmap_yolo/projects/habitat_uncertainty/config.py)
Created customRLConfig with following changes

- Agent: Created a new single agent access manager class.

6. config file for experiments: [YOLOSAM_gaze_rl_skill](https://github.com/AnandSingh-0619/home-robot/blob/anand_dev_heatmap_yolo/projects/habitat_uncertainty/config/YOLOSAM_gaze_rl_skill.yaml)

- DDPPO config: Created new customDDPPOconfig with addition of a new boolean key train_detector.

### CODE REVIEW

**Comment by yusufali98** [comment]
Is there a reason that you dont set inference mode (as done for [visual features](https://github.com/AnandSingh-0619/home-robot/blob/anand_dev_heatmap_yolo/projects/habitat_uncertainty/trainers/YOLOSAM_trainer.py#L238)) when [executing yolo_sam inference](https://github.com/AnandSingh-0619/home-robot/blob/anand_dev_heatmap_yolo/projects/habitat_uncertainty/trainers/YOLOSAM_trainer.py#L230-L231) in the trainer init ? But I guess you do it in [subsequent steps](https://github.com/AnandSingh-0619/home-robot/blob/anand_dev_heatmap_yolo/projects/habitat_uncertainty/trainers/YOLOSAM_trainer.py#L334-L335)

**Comment by AnandSingh-0619** [comment]
Not needed to set that as it is called using Ultralytics library predict function

**Comment by yusufali98** [comment]
Okay got it, but I would still suggest following habitat code as closely as possible so just add that to make things as similar to visual_features_caching as possible

Btw, you should be very careful with hardcoding resolution as [here](https://github.com/AnandSingh-0619/home-robot/blob/anand_dev_heatmap_yolo/projects/habitat_uncertainty/trainers/YOLOSAM_trainer.py#L257)

**Comment by yusufali98** [comment]
Also, what is the reason for doing [this](https://github.com/AnandSingh-0619/home-robot/blob/anand_dev_heatmap_yolo/projects/habitat_uncertainty/trainers/YOLOSAM_trainer.py#L243-L245) ?

**Comment by yusufali98** [comment]
[backbone visual encoder](https://github.com/AnandSingh-0619/home-robot/blob/anand_dev_heatmap_yolo/projects/habitat_uncertainty/models/YOLOSAMPointNavResNetPolicy.py#L855-L863) should correspond to what is set in the [config](https://github.com/AnandSingh-0619/home-robot/blob/anand_dev_heatmap_yolo/projects/habitat_uncertainty/config/YOLOSAM_gaze_rl_skill.yaml#L171)

Also I would strongly suggest you to clean up your policy class implementation. Remove all the unnecessary unused classes/functions/if-else conditions such as [this](https://github.com/AnandSingh-0619/home-robot/blob/anand_dev_heatmap_yolo/projects/habitat_uncertainty/models/YOLOSAMPointNavResNetPolicy.py#L500-L612), [this](https://github.com/AnandSingh-0619/home-robot/blob/anand_dev_heatmap_yolo/projects/habitat_uncertainty/models/YOLOSAMPointNavResNetPolicy.py#L1039-L1050), [this](https://github.com/AnandSingh-0619/home-robot/blob/anand_dev_heatmap_yolo/projects/habitat_uncertainty/models/YOLOSAMPointNavResNetPolicy.py#L937-L971)

Basically, since we are trying to debug issues I would suggets you to remove thigns from implementation which you know should not be used. You can add them back later for future experiments

**Comment by yusufali98** [comment]
Please remove [this condition](https://github.com/AnandSingh-0619/home-robot/blob/anand_dev_heatmap_yolo/projects/habitat_uncertainty/models/YOLOSAMPointNavResNetPolicy.py#L464-L467) as it may cause an issue. Can you please re check the logs and confirm that this condition was never executed. For the current experiments, you should raise an error message whenever the encoder tries to perfrom SAM inference as we dont want that

**Comment by yusufali98** [comment]
Why is [all this](https://github.com/AnandSingh-0619/home-robot/blob/anand_dev_heatmap_yolo/projects/habitat_uncertainty/models/single_agent_access_manager.py#L48-L50) hardcoded ? The class name is also the same as the default habitat single_agent_access_manager. Again, this can cause issues later as you wont remember that you hardcoded all these things and even the class name is same as the default habitat class

**Comment by yusufali98** [comment]
What is [this line](https://github.com/AnandSingh-0619/home-robot/blob/anand_dev_heatmap_yolo/projects/habitat_uncertainty/models/YOLOSAMPointNavResNetPolicy.py#L303) for ?

**Comment by AnandSingh-0619** [comment]
> Also, what is the reason for doing [this](https://github.com/AnandSingh-0619/home-robot/blob/anand_dev_heatmap_yolo/projects/habitat_uncertainty/trainers/YOLOSAM_trainer.py#L243-L245) ?

To delete extra keys from the rollout storage.
head_rgb, yolo_object_sensor, etc
These are present in the env observation space but I have deleted them in rollout storage space. This line create another batch just which has reading only from sensors in rollout_storage_space. Need to do this because there is assertion condition to match observation space before inserting in rollout storage.

**Comment by AnandSingh-0619** [comment]
Net Parameter Distribution
                                                          Parameters

    YOLO                                          164660809 (164M)  (Not trainable)
    SAM                                            10130092 (10M)       (Not trainable)
    Visual Encoder                           3034092 (3M)
    Visual Features                          1044992 (1M)
    State Encoder                             5402624 (5M)
    start_receptacles_embedding   672
    Previous action embedding       192

Total Parameters : 9485655

    Critic                 513
    Net                    9482572
    Action Dist        2570

YOLO and SAM parameters are not listed in total params because YOLOPerception class does not inherit nn.Module (it inherits PerceptionModule same as done by DETIC). I tried to change this but it gives an error says self.train() cannot be boolean.
All the other parameters are trainable and listed with Total Params.
image

**Comment by yusufali98** [comment]
I see the latest commit has quite a few changes
Can you specify whats the main thing that removes the errors that you were running into ?

**Comment by AnandSingh-0619** [comment]
Conversion of detector segmenter output to tensor and store them on GPU eliminated the error.
It is implemented in the code [here](https://github.com/AnandSingh-0619/home-robot/blob/f801daa2957d525e53d6c0006cc56c9773d3b18b/projects/habitat_uncertainty/trainers/YOLOSAM_trainer.py#L230C1-L232C179)

**Comment by yusufali98** [comment]
So earlier `masks` was a CPU numpy array ? And now you are explicitly moving it to the GPU ?

**Comment by yusufali98** [comment]
Also the same "task killed" error shows up in the other yolo_detector only experiments without SAM right ?

**Comment by yusufali98** [comment]
Also, it is not clear to me how does this fix the GPU OOM error in the detector+SAM experiments ? Can you please clarify ?
